### DISCUSSION TITLE
Instantant Solar Radiation from GFS Data

### DISCUSSION THREAD

**Discussion by pedrocorreaa** [open-meteo/open-meteo#917] [Q&A] [answered]
I have noticed that the Open-Meteo API provides Instant Radiation Forecast, including data from the GFS model. However, the GFS output for radiation is averaged over the past N hours (N varies depending on the hour of the GRIB file). [For example](https://www.nco.ncep.noaa.gov/pmb/products/gfs/gfs.t00z.pgrb2.0p25.f003.shtml), in the f003 file, radiation variables are described as the average between hours 0 and 3).

Is there any documentation on how the Instant Radiation is derived from GFS output data?

By the way, great work! The Open-Meteo API is excellent.

**Comment by patrick-zippenfenig** [comment, marked as answer]
Hi,

yes, it is correct that GFS and other weather models return solar fluxes as averages over forecast time. 

1. Open-Meteo first deaverages over the entire forecast time. E.g. given averages `mean 0-1 hours`,  `mean 0-2 hours`,  `mean 0-3 hours`,  `mean 0-4 hours`. This step is done by the [Deaverager](https://github.com/open-meteo/open-meteo/blob/98c62a68fb9ede948c4e1c2ceb0c3fe54070456e/Sources/App/Helper/Writer/GenericVariableHandle.swift#L264).
2. In case the average switches reference point (e.g. forecast hour 24 instead of 0: `mean 24-25 hours`,  `mean 24-26 hours`) this is correctly picked up. Some models were doing it, but I am not sure anymore which ones.
3. If GFS switches from 1 hourly to 3 hourly resolution, intermediate values are interpolated using [clearness index `kt` and an analytical solution over the solar zenith angle](https://github.com/open-meteo/open-meteo/blob/98c62a68fb9ede948c4e1c2ceb0c3fe54070456e/Sources/App/Helper/InterpolationInplace.swift#L217). As an [SPA, NREL is used](https://github.com/open-meteo/open-meteo/blob/98c62a68fb9ede948c4e1c2ceb0c3fe54070456e/Sources/App/Helper/Solar/SolarPositionAlgorithm.swift#L4).

The following steps are performed to get a consistent time series with 1-hour backwards averaged solar radiation. To convert 1-hour backwards to instantaneous radiation, an analytical solution over the zenith angle is used. Code [here](https://github.com/open-meteo/open-meteo/blob/98c62a68fb9ede948c4e1c2ceb0c3fe54070456e/Sources/App/Helper/Solar/Zensun.swift#L284). This assumes that the clearness index `kt` is the same.

I have compared all routines to reference data and there are plenty of unit tests. Hopefully everything is correct :D

**Comment by pedrocorreaa** [comment]
Hi Patrick,

Thank you very much for your response. I will review the code of the Deaverager.

I still have a question regarding the averaging of GFS data. When you mention, for instance, `mean 0-2 hours`, do you mean the average over time between 0:00 and 2:59, or is it between 0:00 and 2:00?

Regards,

**Reply by patrick-zippenfenig** [reply]
It would be the average between 0:00 and 2:00

**Reply by pedrocorreaa** [reply]
thank you!
